runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java

```diff
@@ -157,19 +163,30 @@ public Timer timer(String timerId) {
                 "Access to timers not supported in Splittable DoFn");
           }
         });
-    // TODO: verify that if there was a failed tryClaim() call, then cont.shouldResume() is false.
-    // Currently we can't verify this because there are no hooks into tryClaim().
-    // See https://issues.apache.org/jira/browse/BEAM-2607
     processContext.cancelScheduledCheckpoint();
     @Nullable KV<RestrictionT, Instant> residual = processContext.getTakenCheckpoint();
     if (cont.shouldResume()) {
+      checkState(
+          !processContext.hasClaimFailed,
+          "After tryClaim() returned false, @ProcessElement must return stop(), "
+              + "but returned resume()");
       if (residual == null) {
         // No checkpoint had been taken by the runner while the ProcessElement call ran, however
         // the call says that not the whole restriction has been processed. So we need to take
         // a checkpoint now: checkpoint() guarantees that the primary restriction describes exactly
         // the work that was done in the current ProcessElement call, and returns a residual
         // restriction that describes exactly the work that wasn't done in the current call.
-        residual = checkNotNull(processContext.takeCheckpointNow());
+        if (processContext.numClaimedBlocks > 0) {
+          residual = checkNotNull(processContext.takeCheckpointNow());
+          tracker.checkDone();
+        } else {
+          // The call returned resume() without trying to claim any blocks, i.e. it is unaware
+          // of any work to be done at the moment, but more might emerge later. In this case,
+          // we must simply reschedule the original restriction - checkpointing a tracker that
+          // hasn't claimed any work is not allowed.
+          residual = KV.of(tracker.currentRestriction(), processContext.getLastReportedWatermark());
```

Comment:
<!--new_thread; commit:b04f5ed31c0be1f231dd7c1613ea50d31616b127; resolved:0-->
Why not just fail ? This might result in an infinite scheduling loop due to a bug, no ?

Reply:
Clarified in a comment.